Webpack 5 resolution issues

<!--
  Please describe your bug clearly and concisely.
  Include a screenshot or a reproduction to illustrate
  the bug and how to reproduce it.
  Please also leave a note on what version of urql you're using
  and include a list of exchanges (if any).
-->

Please update your package.json correctly for Webapck Resolution in mjs/esm

```
[0] ModuleNotFoundError: Module not found: Error: Can't resolve 'graphql/language/visitor' in '/Volumes/lulu_dev/ecom-web-app/node_modules/@urql/core/dist'
[0] Did you mean 'visitor.mjs'?
[0] BREAKING CHANGE: The request 'graphql/language/visitor' failed to resolve only because it was resolved as fully specified
[0] (probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
[0] The extension in the request is mandatory for it to be fully specified.
[0] Add the extension to the request.
```

urql version & exchanges: latest

**Steps to reproduce**

build on webpack v5

--- a/.changeset/thin-rings-tell.md
+++ b/.changeset/thin-rings-tell.md
@@ -1,4 +1,5 @@
 ---
+'@urql/preact': patch
 '@urql/exchange-execute': patch
 '@urql/exchange-graphcache': patch
 '@urql/exchange-populate': patch

--- a/.changeset/thin-rings-tell.md
+++ b/.changeset/thin-rings-tell.md
@@ -1,0 +1,9 @@
+---
+'@urql/exchange-execute': patch
+'@urql/exchange-graphcache': patch
+'@urql/exchange-populate': patch
+'@urql/core': patch
+'@urql/introspection': patch
+---
+
+Add missing `.mjs` extension to all imports from `graphql` to fix Webpack 5 builds, which require extension-specific import paths for ESM bundles and packages.

--- a/.changeset/thin-rings-tell.md
+++ b/.changeset/thin-rings-tell.md
@@ -7,4 +7,4 @@
 '@urql/introspection': patch
 ---
 
-Add missing `.mjs` extension to all imports from `graphql` to fix Webpack 5 builds, which require extension-specific import paths for ESM bundles and packages.
+Add missing `.mjs` extension to all imports from `graphql` to fix Webpack 5 builds, which require extension-specific import paths for ESM bundles and packages. **This change allows you to safely upgrade to Webpack 5.**

--- a/package.json
+++ b/package.json
@@ -60,7 +60,7 @@
     "@typescript-eslint/eslint-plugin": "^4.1.0",
     "@typescript-eslint/parser": "^4.1.0",
     "babel-plugin-closure-elimination": "^1.3.2",
-    "babel-plugin-modular-graphql": "0.1.3",
+    "babel-plugin-modular-graphql": "1.0.0",
     "babel-plugin-transform-async-to-promises": "^0.8.15",
     "dotenv": "^8.2.0",
     "eslint": "^7.8.1",

--- a/packages/preact-urql/src/hooks/index.ts
+++ b/packages/preact-urql/src/hooks/index.ts
@@ -1,8 +1,3 @@
-export {
-  useQuery,
-  UseQueryArgs,
-  UseQueryResponse,
-  UseQueryState,
-} from './useQuery';
+export * from './useQuery';
 export * from './useMutation';
 export * from './useSubscription';

--- a/scripts/rollup/cleanup-plugin.js
+++ b/scripts/rollup/cleanup-plugin.js
@@ -1,5 +1,6 @@
 import { transformSync as transform } from '@babel/core';
 import { createFilter } from '@rollup/pluginutils';
+import babelPluginModularGraphQL from 'babel-plugin-modular-graphql';
 
 function removeEmptyImports({ types: t }) {
   return {
@@ -36,7 +37,10 @@ function cleanup(opts = {}) {
       }
 
       return transform(code, {
-        plugins: [removeEmptyImports],
+        plugins: [
+          [babelPluginModularGraphQL, { extension: opts.extension }],
+          removeEmptyImports
+        ],
         babelrc: false
       });
     }

--- a/scripts/rollup/config.js
+++ b/scripts/rollup/config.js
@@ -1,9 +1,9 @@
 import genPackageJson from 'rollup-plugin-generate-package-json';
 import { relative, join, dirname, basename } from 'path';
-import { makePlugins } from './plugins';
+import { makePlugins, makeOutputPlugins } from './plugins';
 import * as settings from './settings';
 
-const plugins = makePlugins({ isProduction: false });
+const plugins = makePlugins();
 
 const input = settings.sources.reduce((acc, source) => {
   acc[source.name] = source.source;
@@ -34,47 +34,49 @@ const input = settings.sources.reduce((acc, source) => {
   return acc;
 }, {});
 
-const config = {
+const output = ({ format, isProduction }) => {
+  if (typeof isProduction !== 'boolean')
+    throw new Error('Invalid option `isProduction` at output({ ... })');
+  if (format !== 'cjs' && format !== 'esm')
+    throw new Error('Invalid option `format` at output({ ... })');
+
+  const extension = format === 'esm'
+    ? (settings.hasReact ? '.es.js' : '.mjs')
+    : '.js';
+
+  return {
+    chunkFileNames: '[hash]' + extension,
+    entryFileNames: '[name]' + extension,
+    dir: './dist',
+    exports: 'named',
+    externalLiveBindings: false,
+    sourcemap: true,
+    esModule: false,
+    indent: false,
+    freeze: false,
+    strict: false,
+    format,
+    plugins: makeOutputPlugins({
+      isProduction,
+      extension,
+    })
+  };
+};
+
+export default {
   input,
   external: settings.isExternal,
   onwarn() {},
+  plugins,
+  output: [
+    output({ format: 'cjs', isProduction: false }),
+    output({ format: 'esm', isProduction: false }),
+    !settings.isCI && output({ format: 'cjs', isProduction: true }),
+    !settings.isCI && output({ format: 'esm', isProduction: true }),
+  ].filter(Boolean),
   treeshake: {
     unknownGlobalSideEffects: false,
     tryCatchDeoptimization: false,
     moduleSideEffects: false
   }
 };
-
-const output = (format = 'cjs', ext = '.js') => ({
-  chunkFileNames: '[hash]' + ext,
-  entryFileNames: '[name]' + ext,
-  dir: './dist',
-  exports: 'named',
-  externalLiveBindings: false,
-  sourcemap: true,
-  esModule: false,
-  indent: false,
-  freeze: false,
-  strict: false,
-  format,
-});
-
-export default [
-  {
-    ...config,
-    shimMissingExports: true,
-    plugins,
-    output: [
-      output('cjs', '.js'),
-      output('esm', settings.hasReact ? '.es.js' : '.mjs'),
-    ],
-  },
-  !settings.isCI && {
-    ...config,
-    plugins: makePlugins({ isProduction: true }),
-    output: [
-      output('cjs', '.min.js'),
-      output('esm', settings.hasReact ? '.min.es.js' : '.min.mjs'),
-    ],
-  },
-].filter(Boolean);

--- a/scripts/rollup/config.js
+++ b/scripts/rollup/config.js
@@ -58,7 +58,7 @@ const output = ({ format, isProduction }) => {
     format,
     plugins: makeOutputPlugins({
       isProduction,
-      extension,
+      extension: format === 'esm' ? '.mjs' : '.js',
     })
   };
 };

--- a/scripts/rollup/plugins.js
+++ b/scripts/rollup/plugins.js
@@ -18,7 +18,7 @@ import babelPluginTransformDebugTarget from '../babel/transform-debug-target';
 
 import * as settings from './settings';
 
-export const makePlugins = ({ isProduction } = {}) => [
+export const makePlugins = () => [
   resolve({
     dedupe: settings.externalModules,
     extensions: ['.js', '.jsx', '.ts', '.tsx'],
@@ -49,7 +49,7 @@ export const makePlugins = ({ isProduction } = {}) => [
         compilerOptions: {
           sourceMap: true,
           noEmit: false,
-          declaration: !isProduction,
+          declaration: true,
           declarationDir: settings.types,
           target: 'esnext',
         },
@@ -75,7 +75,6 @@ export const makePlugins = ({ isProduction } = {}) => [
       babelPluginTransformDebugTarget,
       babelPluginTransformPipe,
       babelPluginTransformInvariant,
-      'babel-plugin-modular-graphql',
       'babel-plugin-closure-elimination',
       '@babel/plugin-transform-object-assign',
       settings.hasReact && ['@babel/plugin-transform-react-jsx', {
@@ -93,20 +92,30 @@ export const makePlugins = ({ isProduction } = {}) => [
       }]
     ].filter(Boolean)
   }),
-  isProduction && replace({
-    'process.env.NODE_ENV': JSON.stringify('production')
-  }),
-  !settings.mayReexport && compiler({
-    formatting: 'PRETTY_PRINT',
-    compilation_level: 'SIMPLE_OPTIMIZATIONS'
-  }),
-  cleanup(),
-  isProduction ? terserMinified : terserPretty,
-  isProduction && settings.isAnalyze && visualizer({
-    filename: path.resolve(settings.cwd, 'node_modules/.cache/analyze.html'),
-    sourcemap: true,
-  }),
-].filter(Boolean);
+];
+
+export const makeOutputPlugins = ({ isProduction, extension }) => {
+  if (typeof isProduction !== 'boolean')
+    throw new Error('Missing option `isProduction` on makeOutputPlugins({ ... })');
+  if (extension !== '.mjs' && extension !== '.js')
+    throw new Error('Missing option `extension` on makeOutputPlugins({ ... })');
+
+  return [
+    isProduction && replace({
+      'process.env.NODE_ENV': JSON.stringify('production')
+    }),
+    !settings.mayReexport && compiler({
+      formatting: 'PRETTY_PRINT',
+      compilation_level: 'SIMPLE_OPTIMIZATIONS'
+    }),
+    cleanup({ extension }),
+    isProduction ? terserMinified : terserPretty,
+    isProduction && settings.isAnalyze && visualizer({
+      filename: path.resolve(settings.cwd, 'node_modules/.cache/analyze.html'),
+      sourcemap: true,
+    }),
+  ].filter(Boolean);
+};
 
 const terserPretty = terser({
   warnings: true,

